src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -111,26 +121,268 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
+                {
+                    return new ConcatNCollectionIterator<TSource>(this, nextCollection, 2);
+                }
+                return new ConcatNEnumerableIterator<TSource>(this, next, 2);
+            }
+
+            internal void CopyTo(TSource[] array, int arrayIndex)
+            {
+                Debug.Assert(array != null);
+                Debug.Assert(array.Length - arrayIndex >= Count);
+
+                _first.CopyTo(array, 0);
+                _second.CopyTo(array, _first.Count);
+            }
+
+            internal override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                switch (index)
+                {
+                    case 0: return _first;
+                    case 1: return _second;
+                    default: return null;
+                }
+            }
+
+            public override TSource[] ToArray()
+            {
+                int firstCount = _first.Count; // Cache an interface method call
+                int totalCount = checked(firstCount + _second.Count);
+
+                if (totalCount == 0)
+                {
+                    return Array.Empty<TSource>();
+                }
+
+                var result = new TSource[totalCount];
+
+                _first.CopyTo(result, 0);
+                _second.CopyTo(result, firstCount);
+
+                return result;
+            }
+
+            public override int GetCount(bool onlyIfCheap) => Count; // Getting the count is always cheap.
+        }
+
+        private sealed class ConcatNCollectionIterator<TSource> : ConcatIterator<TSource>
+        {
+            private readonly ConcatIterator<TSource> _previous;
+            private readonly ICollection<TSource> _next;
+            private readonly int _nextIndex;
+
+            internal ConcatNCollectionIterator(ConcatIterator<TSource> previous, ICollection<TSource> next, int nextIndex)
+            {
+                Debug.Assert(previous != null);
+                Debug.Assert(previous is Concat2CollectionIterator<TSource> || previous is ConcatNCollectionIterator<TSource>);
+                Debug.Assert(next != null);
+                Debug.Assert(nextIndex >= 2);
+
+                _previous = previous;
+                _next = next;
+                _nextIndex = nextIndex;
+            }
+
+            private int Count
+            {
+                get
+                {
+                    // Walk the linked list of Concat{2,N}CollectionIterators and call .Count
+                    // on each of the collections.
+                    // Note that we start from the last collection and make our way to the first,
+                    // but the cumulative count will be the same either way.
+                    
+                    int totalCount = _next.Count;
+                    ConcatIterator<TSource> previous = _previous;
+
+                    // TODO: This can be made more readable once we use C# 7 which supports pattern matching.
```

Comment:
This TODO can be removed.  There are likely many places across the corefx codebase that would benefit from this or that new language feature once that feature is available, and I don't think we need to have a small subset of them annotated with TODOs.